### DISCUSSION TITLE
HTTPS Postal javascript form buttons not working; but they do work under HTTP

### DISCUSSION THREAD

**Discussion by JohnHammell** [postalserver/postal#3113] [Help with using Postal] [answered]
When I access my Postal installation via HTTPS, Postal javascript form buttons such as "Create Route" and "Redeliver message" do not work. They animate after I click them, but they just spin doing nothing for as long as I let it sit.

But when using non-HTTPS those same javascript buttons work immediately as intended.

My connection to Postal is as such:

My browser -> postal.example.com -> HAProxy LXD container -> HAProxy in Postal LXD container -> Postal 127.0.0.1:5000

Any idea what might be causing the Postal javascript buttons to not function when I use postal.example.com via HTTPS?

**Comment by willpower232** [comment]
You would probably need to use your browser developer tools to confirm what was happening. You could also try `postal logs web` to see if there was any information in there.

**Reply by JohnHammell** [reply]
Thanks. Your suggestion so far has led me to uncover part of the issue in the logs. "postal logs web" reported the following:

>exception_class=ActionController::InvalidAuthenticityToken exception_message=HTTP Origin header (https://mail.example.com) didn't match request.base_url (http://mail.example.com)

And that was from the full error message of:

>web-1  | 2024-09-04 00:45:50 +0000 ERROR  POST /org/gc/servers/mail-example-com/routes (422) event=request transaction=cb24190758dfe741cab1 controller=RoutesController action=create format=json method=POST path=/org/gc/servers/mail-example-com/routes request_id=a4203b44-6585-4c9a-8a97-24416f7b1465 ip_address=127.0.0.1 status=422 db_runtime=18.93081295490265 exception_class=ActionController::InvalidAuthenticityToken exception_message=HTTP Origin header (https://mail.example.com) didn't match request.base_url (http://mail.example.com) exception_backtrace=/usr/local/bundle/gems/actionpack-7.0.8.1/lib/action_controller/metal/request_forgery_protection.rb:253:in `handle_unverified_request'\n/usr/local/bundle/gems/actionpack-7.0.8.1/lib/action_controller/metal/request_forgery_protection.rb:286:in `handle_unverified_request'\n/usr/local/bundle/gems/actionpack-7.0.8.1/lib/action_controller/metal/request_forgery_protection.rb:275:in `verify_authenticity_token'\n/usr/local/bundle/gems/activesupport-7.0.8.1/lib/active_support/callbacks.rb:400:in `block in make_lambda'

So the "HTTP Origin header" being HTTPS does not match the "request.base_url" which uses HTTP.

Based on that, it appears that Postal is listening to HTTP and not HTTPS on port 5000, so as far as I can tell it's Postal requiring or configured with the "request.base_url" being HTTP.

In postal.yml there is a config for "web_protocol: https" which seemed it may be the way to config Postal to use HTTPS or HTTP.  I tried using that config with "https" and "http" but it does not make a difference as they both have the same resulting error message above.

I also simply tried treating the proxy to Postal, 127.0.0.1:5000 as HTTPS, but that would not work at all. No server was available on that port using the HTTPS protocol, so only HTTP for the actual internal Postal server as far as I can tell.

I also tried the combinations of HTTP(s) to HTTP(s) to Postal(HTTP) on both HAProxy servers (the HAProxy that directs traffic for the shared IP and the separate HAProxy server which replaced Caddy in the Postal LXD container). Everything appears to be viewable in both HTTPS & HTTP, but the javascript form buttons do not work in HTTPS.

Could it be some javascript files hard coded to HTTP causing this? Not really sure where to check for that.

Also, another issue is with logging in, something I forgot to mention in my initial post. When I login via HTTPS I receive an error message stating: 

>The change you wanted was rejected.
Maybe you tried to change something you didn't have access to.
If you are the application owner check the logs for more information.

But that does not happen when I login using HTTP.

Seems to be the "HTTP Origin header (https://mail.example.com) didn't match request.base_url (http://mail.example.com)" causing these issues, but I cannot seem to figure out how to get Postal to use HTTPS for the request.base_url since it is requiring them to match.

**Comment by JohnHammell** [comment, marked as answer]
The solution to these issues for me was to add the following into my main HAProxy LXD container (not the HAProxy which replaced Caddy in the Postal LXD container):

>http-request set-header X-Forwarded-Proto https if { hdr_beg(host) -i mail. }

That sets the X-Forwarded-Proto header to be https when the url accessed is mail.example.com, or, rather, begins with "mail." Postal seems to need this in order to work properly under HTTPS, at least with the way I have it setup.

@willpower232 you're a bit of a time traveler as you seem to have helped solve this riddle with a comment you posted in a semi-related support issue a year ago. https://github.com/orgs/postalserver/discussions/2574#discussioncomment-6753352  I just stumbled upon that old post by you, decided to give it a try and it seems to have worked out.

Thanks for your help with this and for all the help you provide in these forums.
